core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlErrorStrategy.scala

```diff
@@ -0,0 +1,963 @@
+package com.databricks.labs.remorph.parsers.tsql
+
+import com.databricks.labs.remorph.parsers.SqlErrorStrategy
+import com.databricks.labs.remorph.parsers.tsql.TSqlParser._
+import org.antlr.v4.runtime._
+import org.antlr.v4.runtime.misc.IntervalSet
+
+import scala.collection.convert.ImplicitConversions.`collection AsScalaIterable`
+import scala.collection.mutable.ListBuffer
+
+/**
+ * Custom error strategy for SQL parsing <p> While we do not do anything super special here, we wish to override a
+ * couple of the message generating methods and the token insert and delete messages, which do not create an exception
+ * and don't allow us to create an error message in context. Additionally, we can now implement i18n, should that ever
+ * become necessary.</p>
+ *
+ * <p>At the moment, we require valid SQL as input to the conversion process, but if we ever change that strategy, then
+ * we can implement custom recovery steps here based upon context, though there is no improvement on the sync()
+ * method.</p>
+ */
+class TSqlErrorStrategy extends SqlErrorStrategy {
+
+  /**
+   * Generate a message for the error.
+   *
+   * The exception contains a stack trace, from which we can construct a more informative error message than just
+   * mismatched input and a huge list of things we were looking for.
+   *
+   * @param e
+   *   the RecognitionException
+   * @return
+   *   the error message
+   */
+  override protected def generateMessage(recognizer: Parser, e: RecognitionException): String = {
+    val messages = new ListBuffer[String]()
+
+    val stringBuilder = new StringBuilder()
+
+    // We build the messages by looking at the stack trace of the exception, but if the
+    // rule translation is not found, or it is the same as the previous message, we skip it,
+    // to avoid repeating the same message multiple times. This is because a recognition error
+    // could be found in a parent rule or a child rule but there is no extra information
+    // provided by being more specific about the rule location. ALos, in some productions
+    // we may be embedded very deeply in the stack trace, so we want to avoid too many contexts
+    // in a message.
+    val stack = e.getStackTrace
+    stack.foreach { traceElement =>
```

Comment:
```suggestion
    val messages = e.getStackTrace.flatMap { traceElement => ...
```
flatMap would be more idiomatic here.